cli/commands/stack/action.go

```diff
@@ -0,0 +1,228 @@
+	stackFile, err := config.ReadStackConfigFile(ctx, opts)
+
+	if err != nil {
+		return errors.New(err)
+	}
+
+	if err := processStackFile(ctx, opts, stackFile); err != nil {
+		return errors.New(err)
+	}
+
+	return nil
+}
+func processStackFile(ctx context.Context, opts *options.TerragruntOptions, stackFile *config.StackConfigFile) error {
+	baseDir := filepath.Join(opts.WorkingDir, stackCacheDir)
+	if err := os.MkdirAll(baseDir, dirPerm); err != nil {
+		return errors.New(fmt.Errorf("failed to create base directory: %w", err))
+	}
+
+	for _, unit := range stackFile.Units {
+		destPath := filepath.Join(baseDir, unit.Path)
+		dest, err := filepath.Abs(destPath)
+
+		if err != nil {
+			return errors.New(fmt.Errorf("failed to get absolute path for destination '%s': %w", dest, err))
+		}
+
+		src := unit.Source
+		// set absolute path for source if it's not an absolute path or URL
+		if !filepath.IsAbs(unit.Source) && !isURL(unit.Source) {
+			src = filepath.Join(opts.WorkingDir, unit.Source)
+			src, err = filepath.Abs(src)
+
+			if err != nil {
+				opts.Logger.Warnf("failed to get absolute path for source '%s': %v", unit.Source, err)
+				src = unit.Source
+			}
+		}
+
+		opts.Logger.Infof("Processing unit: %s (%s) to %s", unit.Name, src, dest)
+
+		client := &getter.Client{
+			Src:             src,
+			Dst:             dest,
+			Mode:            getter.ClientModeAny,
+			Dir:             true,
+			DisableSymlinks: true,
+			Options: []getter.ClientOption{
+				getter.WithInsecure(),
+				getter.WithContext(ctx),
+			},
+		}
+
+		// setting custom getters
+		client.Getters = map[string]getter.Getter{}
+
+		for getterName, getterValue := range getter.Getters {
+			// setting custom getter for file to not use symlinks
+			if getterName == "file" {
+				client.Getters[getterName] = &StacksFileProvider{}
+			} else {
+				client.Getters[getterName] = getterValue
+			}
+		}
+
+		if err := client.Get(); err != nil {
+			return errors.New(err)
+		}
+	}
+
+	return nil
+}
+
+func isURL(str string) bool {
+	if strings.Contains(str, "//") {
+		return true
+	}
+
+	u, err := url.Parse(str)
+
+	return err == nil && u.Scheme != "" && u.Host != ""
+}
+
+type StacksFileProvider struct {
+	client *getter.Client
+}
+
+// Get implements downloading functionality
+func (p *StacksFileProvider) Get(dst string, u *url.URL) error {
+	src := u.Path
+	fi, err := os.Stat(src)
+
+	if err != nil {
+		return errors.New(fmt.Errorf("source path error: %w", err))
+	}
+
+	if fi.IsDir() {
+		return p.copyDir(src, dst)
+	}
+
+	return p.copyFile(src, dst)
+}
+
+// GetFile implements single file download
+func (p *StacksFileProvider) GetFile(dst string, u *url.URL) error {
+	return p.copyFile(u.Path, dst)
+}
+
+// ClientMode determines if we're getting a directory or single file
+func (p *StacksFileProvider) ClientMode(u *url.URL) (getter.ClientMode, error) {
+	fi, err := os.Stat(u.Path)
+	if err != nil {
+		return getter.ClientModeInvalid, errors.New(err)
+	}
+
+	if fi.IsDir() {
```

Comment:
Why are we ever fetching a file? We can just always set it to get a directory, right? Do we recursively fetch the individual files after loading the directory?

Reply:
Are copied files recursively even if we have multiple subdirectories